README.md
# Table of Contents

- [About The Project](#about-the-project)
- [Pros](#pros)
- [Cons](#cons)
- [Usage and Demo](#usage-and-demo)
- [License](#license)
- [Contact](#contact)

## About The Project

A simple way to show ansi colors and use escape codes on Linux and Windows terminals in a c++17 style with std::cout. .

## Pros

- Only one header.
- Multiplatform: Windows, Linux and UNIX.
- No external dependencies(*), like conio, ncurses, etc.

## Cons

- (*)On Windows you need to use Windows.h. The reason of that is a necessary hack with the API function SetConsoleMode(). On the other hand, that header and its dependencies are easily availables on Windows sytems.
- Colors and behaviors like blink text, may differ from system to system (xterm, tty, windows, kterm, etc.). The best way to deal with it is trial and error.
- Only for Windows 10 v1511 or higher.
- Not the best option if you are looking for performance.

## Usage and Demo

As easy as type somethink like <strong>std::cout \<\< ansi::yellow << \"Yellow text";</strong>.

Functions are doxygen style documented. Best way to learn is to have a look at the header file.

Run the demo to watch ansi codes in action. Some demo screenshots (in Windows):

![Demo Screenshot 1](doc/demo-screenshot1.png)
![Demo Screenshot 2](doc/demo-screenshot2.png)
![Demo Screenshot 3](doc/demo-screenshot3.png)

## License

Distributed under the MIT License.

## Contact

Feel free to send me any feedback.

José Puga - josepuga.programacion@gmail.com


CHANGELOG.txt
1.0   - 2020  First Release.

1.0.1 - 2021-02 
	Added operator ==, != in ansi::rgb_t struct.
	Added ansi::pos_t struct (x and y coordinates).
	Added goto_xy(const ansi::post_t&) method.
	ansi::rgb_t values are now uint8_t instead int

1.0.2 - 2021-03
	Added fg_rgb(const uint32_t) and bg_rgb(const uint32_t)
	UTF8 support for Windows.

				


src/ansi.hpp
/*
MIT License

Copyright (c) 2020-2021 José Puga

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

#ifndef ANSI_HPP
#define ANSI_HPP
#define ANSI_HPP_VERSION "1.0.2"
#include <string>
#include <string_view>
#include <vector>

#if defined(_WIN32)
#include <cstdio>
#include <windows.h>
//Some old MinGW/CYGWIN distributions don't define this:
#ifndef ENABLE_VIRTUAL_TERMINAL_PROCESSING
#define ENABLE_VIRTUAL_TERMINAL_PROCESSING 0x0004
#endif
namespace ansi {   //Warning: no error checked.
	static DWORD dwMode1 = 0;
	static auto dummyFoo = GetConsoleMode(GetStdHandle(STD_OUTPUT_HANDLE), &dwMode1);
    static DWORD dwMode2 = dwMode1 |= ENABLE_VIRTUAL_TERMINAL_PROCESSING;
    static auto dummyFoo2 = SetConsoleMode(GetStdHandle(STD_OUTPUT_HANDLE), dwMode2);
    // UTF8 for Windows. TODO: Restore old CP. But ansi must be a class...
    static auto dummyFoo3 = SetConsoleOutputCP( CP_UTF8 );
}
#endif // _WIN32

namespace ansi {
    typedef struct rgb_t {
        uint8_t r {0};
        uint8_t g {0};
        uint8_t b {0};
        bool operator==(const rgb_t& rhs) const { 
                return r == rhs.r && g == rhs.g && b == rhs.b; }
        bool operator!=(const rgb_t& rhs) const {
                return !(*this == rhs); }
    } rgb_t;

    typedef struct pos_t {
        int x {0};
        int y {0};
        bool operator==(const pos_t& rhs) const {
                return x == rhs.x && y == rhs.y; }
        bool operator!=(const pos_t& rhs) const {
                return !(*this == rhs); }
    } pos_t;

    //Standard 16 Foreground colors
    static const std::vector<std::string_view> _fg = {
        "\033[38;5;0m", "\033[38;5;1m", "\033[38;5;2m", "\033[38;5;3m", 
        "\033[38;5;4m", "\033[38;5;5m", "\033[38;5;6m", "\033[38;5;7m", 
        "\033[38;5;8m", "\033[38;5;9m", "\033[38;5;10m", "\033[38;5;11m", 
        "\033[38;5;12m", "\033[38;5;13m", "\033[38;5;14m", "\033[38;5;15m"
    };

    //Standard 16 Background colors
   static const std::vector<std::string_view> _bg = {
        "\033[48;5;0m", "\033[48;5;1m", "\033[48;5;2m", "\033[48;5;3m", 
        "\033[48;5;4m", "\033[48;5;5m", "\033[48;5;6m", "\033[48;5;7m", 
        "\033[48;5;8m", "\033[48;5;9m", "\033[48;5;10m", "\033[48;5;11m", 
        "\033[48;5;12m", "\033[48;5;13m", "\033[48;5;14m", "\033[48;5;15m"
   };

   //Greyscale foreground colors (0-23)
   static const std::vector<std::string_view> _fg_grey = {     
        "\033[38;5;232m", "\033[38;5;233m", "\033[38;5;234m", "\033[38;5;235m", 
        "\033[38;5;236m", "\033[38;5;237m", "\033[38;5;238m", "\033[38;5;239m", 
        "\033[38;5;240m", "\033[38;5;241m", "\033[38;5;242m", "\033[38;5;243m", 
        "\033[38;5;244m", "\033[38;5;245m", "\033[38;5;246m", "\033[38;5;247m", 
        "\033[38;5;248m", "\033[38;5;249m", "\033[38;5;250m", "\033[38;5;251m", 
        "\033[38;5;252m", "\033[38;5;253m", "\033[38;5;254m", "\033[38;5;255m"
    };

   //Greyscale background colors (0-23)
   static const std::vector<std::string_view> _bg_grey = {
        "\033[48;5;232m", "\033[48;5;233m", "\033[48;5;234m", "\033[48;5;235m", 
        "\033[48;5;236m", "\033[48;5;237m", "\033[48;5;238m", "\033[48;5;239m", 
        "\033[48;5;240m", "\033[48;5;241m", "\033[48;5;242m", "\033[48;5;243m", 
        "\033[48;5;244m", "\033[48;5;245m", "\033[48;5;246m", "\033[48;5;247m", 
        "\033[48;5;248m", "\033[48;5;249m", "\033[48;5;250m", "\033[48;5;251m", 
        "\033[48;5;252m", "\033[48;5;253m", "\033[48;5;254m", "\033[48;5;255m"
    };

    //Special attributes.
    const std::string_view reset = "\033[0m";
    const std::string_view bold = "\033[1m";
    const std::string_view underline = "\033[4m";
    const std::string_view blink = "\033[5m";  //Not Win10 compatible
    const std::string_view reverse = "\033[7m";

    //Basic foreground normal colors (optional name).
    const std::string_view black = "\033[30m";
    const std::string_view red = "\033[31m";
    const std::string_view green = "\033[32m";
    const std::string_view yellow = "\033[33m";
    const std::string_view blue = "\033[34m";
    const std::string_view magenta = "\033[35m";
    const std::string_view cyan = "\033[36m";
    const std::string_view white = "\033[37m";

    //Basic foreground normal colors.
    const std::string_view fg_black = ansi::black;
    const std::string_view fg_red = ansi::red;
    const std::string_view fg_green = ansi::green;
    const std::string_view fg_yellow = ansi::yellow;
    const std::string_view fg_blue = ansi::blue;
    const std::string_view fg_magenta = ansi::magenta;
    const std::string_view fg_cyan = ansi::cyan;
    const std::string_view fg_white = ansi::white;

    //Basic foreground bold colors.
    const std::string_view fg_b_black = "\033[90m";
    const std::string_view fg_b_red = "\033[91m";
    const std::string_view fg_b_green = "\033[92m";
    const std::string_view fg_b_yellow = "\033[93m";
    const std::string_view fg_b_blue = "\033[94m";
    const std::string_view fg_b_magenta = "\033[95m";
    const std::string_view fg_b_cyan = "\033[96m";
    const std::string_view fg_b_white = "\033[97m";

    //Basic background normal colors.
    const std::string_view bg_black = "\033[40m";
    const std::string_view bg_red = "\033[41m";
    const std::string_view bg_green = "\033[42m";
    const std::string_view bg_yellow = "\033[43m";
    const std::string_view bg_blue = "\033[44m";
    const std::string_view bg_magenta = "\033[45m";
    const std::string_view bg_cyan = "\033[46m";
    const std::string_view bg_white = "\033[47m";

    //Basic background normal colors.
    const std::string_view bg_b_black = "\033[100m";
    const std::string_view bg_b_red = "\033[101m";
    const std::string_view bg_b_green = "\033[102m";
    const std::string_view bg_b_yellow = "\033[103m";
    const std::string_view bg_b_blue = "\033[104m";
    const std::string_view bg_b_magenta = "\033[105m";
    const std::string_view bg_b_cyan = "\033[106m";
    const std::string_view bg_b_white = "\033[107m";

    //Special actions
	const std::string_view save_position = "\033[s";
	const std::string_view restore_position = "\033[u";
	const std::string_view hide_cursor = "\033[?25l";
	const std::string_view show_cursor = "\033[?25h";

    //Clear creen actions
	const std::string_view clear_all = "\033[2J";
	const std::string_view clear_line = "\033[2K";
	const std::string_view clear_to_top = "\033[1J";
	const std::string_view clear_to_bottom = "\033[0J";
	const std::string_view clear_line_to_begin = "\033[1K";
	const std::string_view clear_line_to_end = "\033[0K";
	
    /**
     * @brief Set RGB foreground color
     * @param rgb_t Red (0-255), Green(0-255), Blue(0-255) struct.
     * @return Ansi sequence.
     * */
    inline const std::string fg_rgb(const ansi::rgb_t& t) {
        return  "\033[38;2;" + std::to_string(t.r) + ";" + std::to_string(t.g) 
            + ";" + std::to_string(t.b) + "m";
    }

    /**
     * @brief Set RGB foreground color
     * @param r Red 0-255
     * @param g Green 0-255
     * @param b Blue 0-255
     * @return Ansi sequence.
     * */
    inline const std::string fg_rgb(const uint8_t r, const uint8_t g, const uint8_t b) {
        return  "\033[38;2;" + std::to_string(r) + ";" + std::to_string(g) + ";"
            + std::to_string(b) + "m";
    }

    /**
     * @brief Set RGB foreground color
     * @param color uint32_t color value.
     * @return Ansi sequence.
     * */
    inline const std::string fg_rgb(const uint32_t c) {
        auto r = (c >> 16) & 0xff;
        auto g = (c >> 8) & 0xff;
        auto b = c & 0xff;
        return  "\033[38;2;" + std::to_string(r) + ";" + std::to_string(g) 
            + ";" + std::to_string(b) + "m";
    }


    /**
     * @brief Set RGB background color
     * @param rgb_t Red (0-255), Green(0-255), Blue(0-255) struct.
     * @return Ansi sequence.
     * */
    inline const std::string bg_rgb(const ansi::rgb_t& t) {
        return  "\033[48;2;" + std::to_string(t.r) + ";" + std::to_string(t.g) +
            ";" + std::to_string(t.b) + "m";
    }

    /**
     * @brief Set RGB background color
     * @param r Red 0-255
     * @param g Green 0-255
     * @param b Blue 0-255
     * @return Ansi sequence.
     * */
    inline const std::string bg_rgb(const uint8_t r, const uint8_t g, const uint8_t b) {
        return  "\033[48;2;" + std::to_string(r) + ";" + std::to_string(g) + ";"
            + std::to_string(b) + "m";
    }

    /**
     * @brief Set RGB background color
     * @param color uint32_t color value.
     * @return Ansi sequence.
     * */
    inline const std::string bg_rgb(const uint32_t c) {
        auto r = (c >> 16) & 0xff;
        auto g = (c >> 8) & 0xff;
        auto b = c & 0xff;
        return  "\033[48;2;" + std::to_string(r) + ";" + std::to_string(g) 
            + ";" + std::to_string(b) + "m";
    }

    /**
     * @brief Set standard foreground color. 
     * @param color Color code 0-15.
     * @return Ansi sequence.
     * */
    inline const std::string_view fg(const int color) {
        return ansi::_fg.at(color);
    }

    /**
     * @brief Set standard background color. 
     * @param color Color code 0-15.
     * @return Ansi sequence.
     * */
    inline const std::string_view bg(const int color) {
        return ansi::_bg.at(color);
    }

    /**
     * @brief Set the foreground color in a grey scale of 24. 0 is the 
     * darkest and 23 the whitest.
     * @param color Color code 0-23
     * @return Ansi sequence.
     * */
    inline const std::string_view fg_grey(const int color) {
        return ansi::_fg_grey.at(color);
    }

    /**
     * @brief Set the background color in a grey scale of 24. 0 is the 
     * darkest and 23 the whitest.
     * @param color Color code 0-23
     * @return Ansi sequence.
     * */    
    inline const std::string_view bg_grey(const int color) {
        return ansi::_bg_grey.at(color);
    }

	/**
     * @brief Moves cursor position n times up.
     * @param n N times to move, 1 by default.
     * @return Ansi sequence.
     * */
    inline const std::string up(const int n = 1) { 
        return "\033[" + std::to_string(n) + "A"; }

	/**
     * @brief Moves cursor position n times down.
     * @param n N times to move, 1 by default.
     * @return Ansi sequence.
     * */
    inline const std::string down(const int n = 1) { 
        return "\033[" + std::to_string(n) + "B"; }

	/**
     * @brief Moves cursor position n times right.
     * @param n N times to move, 1 by default.
     * @return Ansi sequence.
     * */
    inline const std::string right(const int n = 1) { 
        return "\033[" + std::to_string(n) + "C"; }

	/**
     * @brief Moves cursor position n times left.
     * @param n N times to move, 1 by default.
     * @return Ansi sequence.
     * */
    inline const std::string left(const int n = 1) { 
        return "\033[" + std::to_string(n) + "D"; }

	/**
     * @brief Delete n characters from the cursor position to the right.
     * @param n N characters to delete, 1 by default.
     * @return Ansi sequence.
     * */
    inline const std::string delete_char(const int n = 1) { 
        return "\033[" + std::to_string(n) +"Km"; }

	/**
     * @brief Set cursor position.
     * @param x Colum.
     * @param y Row.
     * @return Ansi sequence.
     * */
	inline const std::string goto_xy(const int x, const int y) { 
        return "\033[" + std::to_string(y) + ";" + std::to_string(x) + "f"; }

	/**
     * @brief Set cursor position.
     * @param pos_t Struct with X and Y coordinates.
     * @return Ansi sequence.
     * */
	inline const std::string goto_xy(const ansi::pos_t pos) { 
        return goto_xy(pos.x, pos.y); }

} //namespace

#endif // ANSI_HPP



demo_ansi.cpp
//Demo for ansi.h by José Puga
#include <iomanip> 
#include <iostream>
#include <thread> 
#include <chrono> 
#include <map>
#include <variant>
#include "ansi.hpp"     //Only one header

using namespace std;

void Pause()
{
    cout << ansi::white << ansi::bg_black << "\nPress " << ansi::bold <<
    ansi::underline << "ENTER" << ansi::reset << " to continue the demo...";
    cin.get();
}

void ClearAndReset()
{
    //ansi::reset to turn off special attributes like blink or bold.
    //ansi::clear_all to clear screen
    //Strange behavior sometimes clear_all is not enought in Windows.
    //Is good idea add goto_xy(1,1)
    cout << ansi::reset << ansi::clear_all << ansi::goto_xy(1,1);
}
int main()
{ 
    cout << "Holá ñaña middle·dot";
    Pause();
    return 0;
    //TEST STANDARD 16 FOREGROUNDS COLORS    
    ClearAndReset();
    cout << ansi::right(10) << ansi::underline << 
          "FOREGROUND STANDARD 16 COLORS TEST:" << ansi::reset;
    cout << "\nUsing " << ansi::yellow << "ansi::fg(n)" << 
        ansi::reset << ". Where n = 0-15:\n\n";

    auto maxColums = 8, colum = 0; //Demo with 80 colums
    for (auto i = 0; i < 16; i++) {
        cout << ansi::bg_black << ansi::white << " " << setw(3) << i << "=";
        cout << ansi::fg(i) << "#%@AB";
        colum++;
        if (colum == maxColums) {
            colum = 0;
            cout << "\n";
        }
    }

    //TEST STANDARD 16 BACKGROUNDS COLORS    
    cout << "\n\n\n" << ansi::right(10) << ansi::underline <<
        "BACKGROUND STANDARD 16 COLORS TEST:" << ansi::reset;
    cout << "\nUsing " << ansi::yellow << "ansi::bg(n)" << 
        ansi::reset << ". Where n = 0-15:\n\n";
    colum = 0;
    for (auto i = 0; i < 16; i++) {
        cout << ansi::bg_black << ansi::white << " " << setw(3) << i << "=";
        cout << ansi::bg(i) << "#%@AB";
        colum++;
        if (colum == maxColums) {
            colum = 0;
            cout << "\n";
        }
    }
    Pause();
    ClearAndReset();

    //TEST GREYSCALE FOREGROUND COLORS    
    cout << ansi::right(10) << ansi::underline << "GREYSCALE COLORS TEST:" << 
        ansi::reset;
    cout << "\n\nTest Foreground Greyscale using " << ansi::yellow << 
        "ansi::fg_grey(n)" << ansi::reset << " where n = 0-23:\n\n";
    colum = 0;
    for (auto i = 0; i < 24; i++) {
        cout << ansi::bg_black << ansi::white << " " << setw(3) << i << "=";
        cout << ansi::fg_grey(i) << "#%@AB";
        colum++;
        if (colum == maxColums) {
            colum = 0;
            cout << "\n";
        }
    }

    //TEST GREYSCALE BACKGROUND COLORS    
    cout << ansi::reset;
    cout << "\n\n\nTest Background Greyscale using " << ansi::yellow
        << "ansi::bg_grey(n)" << ansi::reset << " where n = 0-23:\n\n";
    colum = 0;
    for (auto i = 0; i < 24; i++) {
        cout << ansi::bg_black << ansi::white << " " << setw(3) << i << "=";
        cout << ansi::bg_grey(i) << "#%@AB";
        colum++;
        if (colum == maxColums) {
            colum = 0;
            cout << "\n";
        }
    }
    Pause();
    ClearAndReset();

    //TEST COLOR NAMED
    cout << ansi:: goto_xy(1,4) << "The most common and most compatible colors"
        << " are the first 16.\nYou can use they directly with the color name."
        << "\n" <<
        "Like " << ansi::blue << "ansi::blue" << ansi::reset << " or " <<
        ansi::fg_blue << "ansi::fg_blue" << ansi::reset << 
        " for normal foreground blue,\n" << 
        "or " << ansi::fg_b_blue << "ansi::fb_b_blue" << ansi::reset << 
        " for bold foreground blue.\n\n";
    cout << "Same for background colors. Normal like " << ansi::bg_green <<
        "'ansi::bg_green'" << ansi::reset << ".\n" <<
        "And bold version " << ansi::bg_b_green << "ansi::bg_b_green" << 
        ansi::reset << ".\n";
    cout << "ansi_fg::XXX (or ansi::XXX), ansi::fg_b_XXX, ansi::bg_XXX, ansi::"
        << "bg_b_XXX.\nWhere XXX can be: red, green, yellow, blue, "
        << "magenta, cyan, white, black. (8 normal + 8 bold = 16)\n";
    Pause();
    ClearAndReset();

    //TEST RGB COLORS
    cout << "Of course, RBG colors are availables. fg_rgb or bg_rgb:\n" << ansi::yellow << 
        "ansi::fg_rgb(const uint32_t color)\n" <<
        "ansi::fg_rgb(const uint8_t r, const uint8_t g, const uint8_t b)\n" <<
        "ansi::fg_rgb(const ansi::rgb_t& rgb)\n\n" << ansi::reset;

    //Common HTML color names: https://www.w3schools.com/colors/colors_names.asp
    map<string, int> commonsRGB {
        {"Maroon", 0x800000},
        {"Olive", 0x808000},
        {"Aqua", 0x00FF37},
        {"Purple", 0X800080},
        {"Sienna", 0xa0522d},
        {"Teal", 0x008080},
        {"Torquoise", 0x40e0d0},
        {"Khaki", 0xf0e68c},
        {"Wheat", 0xf5deb3},
        {"Indigo", 0x4b0082},
        {"Azure", 0xf0ffff},
        {"Almond",0xffebcd}
    };
    maxColums = 4;
    colum = 0;
    for (auto it = commonsRGB.begin(); it != commonsRGB.end(); it++) {
        cout << ansi::reset << setw(11) << right << it->first << ": " <<
            ansi::fg_rgb(it->second) << "#%@AB";
        colum++;
        if (colum == maxColums) {
            colum = 0;
            cout << "\n";
        }
    }

    cout << "\n";
    colum = 0;
    for (auto it = commonsRGB.begin(); it != commonsRGB.end(); it++) {
        cout << ansi::reset << setw(11) << right << it->first << ": " <<
            ansi::bg_rgb(it->second) << "#%@AB";
        colum++;
        if (colum == maxColums) {
            colum = 0;
            cout << "\n";
        }
    }

    cout << ansi::reset << "\n\nSmooth greyscale with RGB: ";
    for (auto i = 0; i < 255; i+=8) {
        cout << ansi::bg_rgb(i, i, i) << "*";
    }
    cout << "\n";
    Pause();
    ClearAndReset();
    //TEST SAVE/RESTORE CURSOR POSITION
    cout << "\n\nYou can save and restore the cursor position at any time with\n" <<
        ansi::yellow << "ansi::save_position" << ansi::reset << " and " << 
        ansi::yellow << "ansi::restore_position" << ansi::reset << ".\n" <<
        "Now press " << ansi::underline << "ENTER" << ansi::reset << 
        " to save your cursor position.\n" << ansi::save_position;
    cin.get();
    cout << "Press ENTER again to restore the position and write the text " <<
        "'Here we are...'\n\n\n";
    cin.get();
    cout << ansi::restore_position << ansi::blink << "Here we are..." <<
        ansi::reset << " (Press ENTER)";
    cin.get();

    //TEST GOTO_XY AND HIDE CURSOR
    ClearAndReset();
    cout << ansi::goto_xy(1, 10);
    cout << "Two of the most powerful features are\n" << ansi::yellow << 
        "ansi::goto_xy(x, y)" << ansi::reset << " and " << ansi::yellow <<
        "ansi::up(), down(), left() or right()" << ansi::reset << ".\n" <<
        "In this sample I'm using " << ansi::yellow << "ansi::hide_cursor" <<
        ansi::reset << " too.\n\n" << 
        ansi::hide_cursor << ansi::save_position;
    srand(time(0)); // Yes I know!. Rand and Seed C style...
    for (auto i = 0; i < 85; i++) {
        cout << ansi::goto_xy((rand() % 40) + 10, (rand() % 6) + 2) <<
            ansi::bg((rand() % 15)+1) << " ";
        std::this_thread::sleep_for (std::chrono::milliseconds(100));
        cout << flush; //On some linux terminals you need to flush the buffer.
    }

    //TEST SHOW AND MOVE CURSOR
    cout << ansi::show_cursor << ansi::reset;
    for (auto i = 0; i < 2; i++) {
        //NOTE: flush is only necessary on some linux terminals.
        cout << ansi::right(10) << flush;
        std::this_thread::sleep_for (std::chrono::milliseconds(500));
        cout <<  ansi::right(3) << flush;
        std::this_thread::sleep_for (std::chrono::milliseconds(500));
        cout <<  ansi::left() << flush;
        std::this_thread::sleep_for (std::chrono::milliseconds(500));
        cout <<  ansi::left() << flush;
        std::this_thread::sleep_for (std::chrono::milliseconds(500));
        cout <<  ansi::down() << flush;
    }
    cout << ansi::restore_position << ansi::reset << 
        "There are more features, don't forget to read the header file.\nThanks for watching. Jose Puga.";
    cin.get();

    cout << ansi::white << ansi::bg_black << "\n"; //Default console colors fg/bg
    return 0;
}
